### ISSUE TITLE
JSON messages with RFC 3339 / ISO 8601 nanosecond timestamps not partitioned as expected to S3

### ISSUE THREAD

**Issue by gwilym** [pinterest/secor#305]
When I analyse the files placed on S3 by secor I find there are messages being placed incorrectly with regards to date partitioning. Each JSON message has a top-level `timestamp` property which can be in either `2006-01-02T15:04:05.999999999Z` format or, occasionally, `2006-01-02T15:04:05Z` (as in, without nanoseconds, from systems which do not support such resolution).

Note: nanoseconds may not be a culprit here as the title suggests, that's just the majority of values in my dataset.

I've tried this with 0.22 from maven, and master (aee073b6a105f4365f50cabac1248747781a7a9a) from github.

It seems to work fine for the majority of events. However, when I scan the files for the dates of each message, I get results like...

```
# aws s3 ls --recursive s3://BUCKETNAME/secor-test-9/ | tr -s ' ' | cut -d' ' -f4 | sort | xargs -I{} sh -c 'echo {}; aws s3 cp s3://BUCKETNAME/{} - | zcat | jq ".timestamp[0:10]" | sort | uniq -c'
```

```
secor-test-9/athena-integration-1/dt=2017-01-31/1_22_00000000000002657819.gz
  19921 "2017-01-31"
      2 "2017-02-01"

secor-test-9/athena-integration-1/dt=2017-02-01/1_0_00000000000002675808.gz
  19302 "2017-02-01"
      3 "2017-02-02"
      1 "2017-02-03"

secor-test-9/athena-integration-1/dt=2017-02-01/1_11_00000000000002717065.gz
      2 "2017-02-02"
      2 "2017-02-03"
      5 "2017-02-04"
      3 "2017-02-05"
      2 "2017-02-06"

secor-test-9/athena-integration-1/dt=2017-02-01/1_12_00000000000002655329.gz
      3 "2017-01-31"
  33460 "2017-02-01"
      1 "2017-02-02"

secor-test-9/athena-integration-1/dt=2017-02-04/1_3_00000000000002720791.gz
      2 "2017-02-02"
  12869 "2017-02-04"
```

For one specific example above `secor-test-9/athena-integration-1/dt=2017-02-01/1_11_00000000000002717065.gz` the timestamps of each line are

```
"2017-02-02T22:33:03.12533900Z"
"2017-02-02T22:46:12.26030700Z"
"2017-02-03T00:49:04.27147800Z"
"2017-02-03T05:21:52.79127500Z"
"2017-02-04T00:12:43.81382200Z"
"2017-02-04T04:39:16.33926200Z"
"2017-02-04T04:41:33.62567100Z"
"2017-02-04T05:21:55.08586200Z"
"2017-02-04T07:39:26.89454900Z"
"2017-02-05T07:13:58.89362700Z"
"2017-02-05T07:59:05.27761600Z"
"2017-02-05T23:16:37.77798300Z"
"2017-02-06T19:20:29.98021100Z"
"2017-02-06T19:44:06.95325400Z"
```

Note that these nanosecond timestamps are handled correctly though for the vast majority (see: tens of thousands of correctly partitioned events).

My normalised (comments removed etc.) config file is:

```
aws.proxy.http.port=
aws.proxy.isEnabled=false
aws.region=
aws.role=
aws.secret.key=
aws.sse.customer.key=
aws.sse.kms.key=
aws.sse.type=
cloud.service=S3
hive.table.prefix=
kafka.consumer.auto.offset.reset=smallest
kafka.consumer.timeout.ms=10000
kafka.dual.commit.enabled=false
kafka.fetch.message.max.bytes=
kafka.fetch.min.bytes=
kafka.fetch.wait.max.ms=
kafka.offsets.storage=zookeeper
kafka.partition.assignment.strategy=range
kafka.rebalance.backoff.ms=
kafka.rebalance.max.retries=
kafka.seed.broker.host=...
kafka.seed.broker.port=9092
kafka.socket.receive.buffer.bytes=
kafka.zookeeper.path=/kafka-kafka
message.timestamp.id=1
message.timestamp.input.pattern=
message.timestamp.name.separator=
message.timestamp.name=timestamp
message.timestamp.required=true
message.timestamp.type=i64
monitoring.blacklist.topics=
monitoring.prefix=secor
ostrich.port=9999
partitioner.finalizer.delay.seconds=3600
partitioner.granularity.hour=true
partitioner.granularity.minute=false
qubole.api.token=
secor.compression.codec=org.apache.hadoop.io.compress.GzipCodec
secor.consumer.threads=7
secor.enable.qubole=true
secor.file.extension=
secor.file.reader.writer.factory=com.pinterest.secor.io.impl.DelimitedTextFileReaderWriterFactory
secor.generation=1
secor.kafka.group=secor-test-9
secor.kafka.topic_blacklist=
secor.kafka.topic_filter=^athena-((?!production).)*$
secor.kafka.upload_at_minute_mark.topic_filter=
secor.local.log.delete.age.hours=-1
secor.local.path=/tmp/secor-test-9
secor.max.file.age.seconds=120
secor.max.file.size.bytes=20000000
secor.max.message.size.bytes=100000
secor.message.parser.class=com.pinterest.secor.parser.Iso8601MessageParser
secor.message.transformer.class=com.pinterest.secor.transformer.IdentityMessageTransformer
secor.messages.per.second=100000
secor.offsets.per.partition=10000000
secor.offsets.prefix=offset=
secor.parser.timezone=UTC
secor.qubole.timeout.ms=300000
secor.s3.alter.path.date=
secor.s3.alternative.path=
secor.s3.bucket=...
secor.s3.filesystem=s3a
secor.s3.path=secor-test-9
secor.s3.prefix.md5hash=false
secor.topic_partition.forget.seconds=600
secor.upload.manager.class=com.pinterest.secor.uploader.S3UploadManager
secor.upload.minute_mark=0
secor.zookeeper.path=/secor-test-9
statsd.hostport=
tsdb.hostport=
zookeeper.quorum=...
zookeeper.session.timeout.ms=3000
zookeeper.sync.time.ms=200
```

This could be a bug, or it could be a misconfiguration on my behalf, but I've played around with this config for the better part of a day using the available documentation and can't improve it beyond this. Hourly granularity doesn't seem to work either in this setup, but that's another issue I suppose.

**Comment by HenryCaiHaiying** [comment]
Take a look at the code in Iso8601MessageParser to see if it can handles
your two time formats.

**Comment by gwilym** [comment]
While I know the syntax, I'm not familiar enough with the Java ecosystem to tell if it can or not, based on the libraries it's using.

**Comment by HenryCaiHaiying** [comment]
You can write a small java program using the code in Iso860MessageParser to
test your json.

Or add your json with different date formats in Iso8601ParserTest.java to
run the test.

**Comment by gwilym** [comment]
@HenryCaiHaiying I've added the test case in #306 -- the time formats appear to be supported by the parser itself

**Comment by gwilym** [comment]
I guess github interprets "does not fix #n" as "fix #n". @HenryCaiHaiying would you mind re-opening this?

**Comment by gwilym** [comment]
fixed using #307 -- see config sample there
